src/Symfony/Component/HttpClient/Rfc9111HttpClient.php

```diff
@@ -0,0 +1,605 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Symfony\Component\Cache\Adapter\TagAwareAdapterInterface;
+use Symfony\Component\HttpClient\Response\AsyncContext;
+use Symfony\Component\HttpClient\Response\AsyncResponse;
+use Symfony\Component\HttpClient\Response\MockResponse;
+use Symfony\Component\HttpClient\Response\ResponseStream;
+use Symfony\Contracts\HttpClient\ChunkInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+use Symfony\Contracts\Service\ResetInterface;
+
+/**
+ * Adds caching on top of an HTTP client (per RFC 9111).
+ *
+ * Known omissions / partially supported features per RFC 9111:
+ *   1. Range requests:
+ *     - All range requests ("partial content") are passed through and never cached.
+ *   2. stale-while-revalidate:
+ *     - There's no actual "background revalidation" for stale responses, they will
+ *       always be revalidated.
+ *   3. min-fresh, max-stale, only-if-cached:
+ *     - Request directives are not parsed; the client ignores them.
+ *
+ * @see https://www.rfc-editor.org/rfc/rfc9111
+ */
+class Rfc9111HttpClient implements HttpClientInterface, ResetInterface
+{
+    use AsyncDecoratorTrait {
+        stream as asyncStream;
+        AsyncDecoratorTrait::withOptions insteadof HttpClientTrait;
+    }
+    use HttpClientTrait;
+
+    private array $defaultOptions = self::OPTIONS_DEFAULTS;
+
+    public function __construct(
+        HttpClientInterface $client,
+        private readonly TagAwareAdapterInterface $cache,
+        array $defaultOptions = [],
+        private readonly bool $sharedCache = true,
+        private readonly ?int $ttl = null,
+        private readonly array $defaultCachableStatusCodes = [200, 203, 204, 206, 300, 301, 404, 405, 410, 414, 501],
+        private readonly array $conditionallyCachableCodes = [302, 303, 307, 308],
+        private readonly array $cacheableMethods = ['GET', 'HEAD'],
+        private readonly array $unsafeMethods = ['POST', 'PUT', 'DELETE', 'PATCH'],
```

Comment:
Do we need this, or can we say "all methods which are not cacheable, are considered unsafe?"

Reply:
No. Unsafe methods are methods that can change the server's state or data. Per RFC 9111, they should trigger a cache invalidation when called (which the current implementation do).

So, if eg. `HEAD` is not configured as a cacheable method by the user, it's still not an unsafe method and shouldn't trigger a cache invalidation when called.

Reply:
We could put as a class const, but I've came across APIs that proposed `POST` endpoints to let their users express complex search queries that wouldn't fit in a regular `GET` endpoint with url parameters, so I wanted to let the users be able to configure that.